### ISSUE TITLE
Odd lambda type inference for functools.reduce with initial values involving literal zeros

### ISSUE THREAD

**Issue by kltws** [microsoft/pyright#6031]
I expected the following to type check correctly
```python
from functools import reduce
from typing import Iterable

def calculate_average(values: Iterable[float]) -> float:
    total, num_of_values = reduce(
        lambda total_and_count, value: (total_and_count[0] + value, total_and_count[1] + 1),
        values,
        (0, 0)
    )
    return total / num_of_values
```
But Pylance (v2023.9.21) reports the following error on the lambda
```
Argument of type "(total_and_count: tuple[Literal[0], Literal[0]] | tuple[float, Literal[1]] | tuple[float, Literal[1, 2]], value: float) -> tuple[float, Literal[1, 2, 3]]" cannot be assigned to parameter "function" of type "(_T@reduce, _S@reduce) -> _T@reduce" in function "reduce"
  Type "(total_and_count: tuple[Literal[0], Literal[0]] | tuple[float, Literal[1]] | tuple[float, Literal[1, 2]], value: float) -> tuple[float, Literal[1, 2, 3]]" cannot be assigned to type "(tuple[Literal[0], Literal[0]] | tuple[float, Literal[1]] | tuple[float, Literal[1, 2]], float) -> (tuple[Literal[0], Literal[0]] | tuple[float, Literal[1]] | tuple[float, Literal[1, 2]])"
    Function return type "tuple[float, Literal[1, 2, 3]]" is incompatible with type "tuple[Literal[0], Literal[0]] | tuple[float, Literal[1]] | tuple[float, Literal[1, 2]]"
      Type "tuple[float, Literal[1, 2, 3]]" cannot be assigned to type "tuple[Literal[0], Literal[0]] | tuple[float, Literal[1]] | tuple[float, Literal[1, 2]]"
        "tuple[float, Literal[1, 2, 3]]" is incompatible with "tuple[Literal[0], Literal[0]]"
          Tuple entry 1 is incorrect type
        "tuple[float, Literal[1, 2, 3]]" is incompatible with "tuple[float, Literal[1]]"
          Tuple entry 2 is incorrect type
        "tuple[float, Literal[1, 2, 3]]" is incompatible with "tuple[float, Literal[1, 2]]"
  ...
```
Casting the second zero to an int ```(0, cast(int, 0))``` or explicitly typing the lambda as ```Callable[[tuple[float, int], float], tuple[float, int]]``` works, but feels like should be avoidable.

**Comment by erictraut** [comment]
This is a case where the lambda's type cannot be correctly inferred from the available context. Lambdas are tricky because their parameter types must always be inferred (because the Python syntax doesn't support explicit type annotations for lambda parameters). 

This particular case is especially tricky because the constraint solver is dealing with tuple types, and it's not able to unify `tuple[int, float]` and `tuple[float, float]`.

I'm not surprised to see that mypy also emits an error in this case — for the same reason.

I'll take a closer look at this to see if there's a way to improve pyright's constraint solver to handle more sophisticated unification of tuple types, but I'm not making any promises at this point. If this adds too much complexity or has other negative side effects, I will probably close this as a "won't fix".

In the meantime, you can work around the issue by specifying the initial value as `(0.0, 0.0)` rather than `(0, 0)`. With this small change, the code type checks in both pyright and mypy.

**Comment by kltws** [comment]
```python
def concatenate_and_count(values: Iterable[str])  -> tuple[str, int]:
    concatenated, num_of_values = reduce(
        lambda cummulative, value: (cummulative[0] + value, cummulative[1] + 1),
        values,
        ('', 0)
    )
    return concatenated, num_of_values
```
(Changed the example to one with `tuple[str, int]`, but as before, the literal `0` in the tuple and the `+ 1` in the lambda causes  the type inference to the more restrictive `tuple[str, Literal[0, 1, 2]]`)

It seems like there is some special casing for addition with literal ints that is causing this particular type inferencing? 
As far as I can find there is no special-casing for literal ints on `int.__add__` in the typeshed stubs so I'm assuming it's in pyright. I'm sure there are cases where that would be useful though in this case it becomes unfortunately prohibitive.

If it does turn out to be too disruptive to try to handle this, I'm fine with that since there are ways to circumvent it (for now I've declared and type hinted the lambda outside of reduce, though in my actual use-case I'm reducing a much longer tuple which meant a much more cumbersome type hint I was hoping to avoid)

**Comment by erictraut** [comment]
This problem was caused by pyright's "literal math" inference feature. For simple operations (like addition) that involve literal `int` or `str` operands, pyright infers the resulting type as a literal. For example:

```python
x = 1 + 2
reveal_type(x) # Literal[3]
```

When operations are performed within a loop, pyright no longer performs "literal math" inference because it's no longer safe to do so. It turns out that it's probably also not a good idea to perform literal math for lambdas because they can be called repeatedly (such as in your two code samples above). I've therefore disabled literal math inference for operations in lambdas. This eliminates the false positive errors in both of your code samples above.

This fix will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.330, which I just published. It will also be included in an upcoming release of pylance.
